crates/bevy_ecs/src/component.rs

```diff
@@ -953,6 +997,283 @@ impl Components {
+        }
+
+        // Propagate the new required components up the chain to all components that require the requiree.
+
+        if let Some(required_by) = self.get_required_by(requiree).cloned() {
+            for &required_by_id in required_by.iter() {
+                // SAFETY: The component is in the list of required components, so it must exist already.
+                let required_components = unsafe {
+                    self.get_required_components_mut(required_by_id)
+                        .debug_checked_unwrap()
+                };
+
+                // Register the original required component for the requiree.
+                // The inheritance depth is `1` since this is a component required by the original requiree.
+                required_components.register_by_id(required, constructor, 1);
+
+                for (component_id, component) in recursive_requires.iter() {
+                    // Register the required component.
+                    // The inheritance depth is increased by `1` since this is a component required by the original required component.
+                    // SAFETY: Component ID and constructor match the ones on the original requiree.
+                    //         The original requiree is responsible for making sure the registration is safe.
+                    unsafe {
+                        required_components.register_dynamic(
+                            *component_id,
+                            component.constructor.clone(),
+                            component.inheritance_depth + 1,
+                        );
+                    };
+                }
+            }
+        }
+    }
+
+    // NOTE: This should maybe be private, but it is currently public so that `bevy_ecs_macros` can use it.
+    //       We can't directly move this there either, because this uses `Components::get_required_by_mut`,
+    //       which is private, and could be equally risky to expose to users.
+    /// Registers the given component `R` as a [required component] for `T`,
+    /// and adds `T` to the list of requirees for `R`.
+    ///
+    /// The given `inheritance_depth` determines how many levels of inheritance deep the requirement is.
+    /// A direct requirement has a depth of `0`, and each level of inheritance increases the depth by `1`.
+    /// Lower depths are more specific requirements, and can override existing less specific registrations.
+    ///
+    /// This method does *not* recursively register required components for components required by `R`,
+    /// nor does it register them for components that require `T`.
+    ///
+    /// Only use this method if you know what you are doing. In most cases, you should instead use [`World::register_required_components`],
+    /// or the equivalent method in `bevy_app::App`.
+    ///
+    /// [required component]: Component#required-components
+    pub fn register_required_components_manual<T: Component, R: Component>(
```

Comment:
Comment on lines 1185-1187:
```rust
// NOTE: This should maybe be private, but it is currently public so that `bevy_ecs_macros` can use it.
//       We can't directly move this there either, because this uses `Components::get_required_by_mut`,
//       which is private, and could be equally risky to expose to users.
```
This is kinda scuffed, but I think it's fine since this is pretty clearly not intended to be a user-facing method based on the name and docs. Any thoughts?

Reply:
Yeah that's fine. Normally I'd do `manually_register_required_componnets`, but that'll break alphabetical sorting.